Add ExMachina.EctoPolymorphicEmbed

atdrendel opened the pull request:
This is an attempt to make [ExMachina](https://github.com/beam-community/ex_machina) work with [polymorphic_embed](https://github.com/mathieuprog/polymorphic_embed) following the lead of this [comment](https://github.com/beam-community/ex_machina/issues/407#issuecomment-2132656371).

atdrendel committed:
Add ExMachina.EctoPolymorphicEmbed

westmark commented:
I was using a local version of the same, and after updating my packages (unsure which caused it) I had to change 

```ex
&match?({:parameterized, PolymorphicEmbed, _options}, schema.__schema__(:type, &1))
```

to 

```ex
&match?({:parameterized, {PolymorphicEmbed, _options}}, schema.__schema__(:type, &1))
```

for the casting to work

teamon commented on lib/ex_machina/ecto_polymorphic_embed_strategy.ex:
```diff
@@ -0,0 +1,155 @@
+      schema.__schema__(:fields),
+      &match?({:parameterized, PolymorphicEmbed, _options}, schema.__schema__(:type, &1))
+    )
+  end
```
Need to account for array of polymorphic embeds:
```ex
  defp schema_polymorphic_embeds(schema) do
    Enum.filter(schema.__schema__(:fields), fn field ->
      case schema.__schema__(:type, field) do
        {:parameterized, {PolymorphicEmbed, _options}} -> true
        {:array, {:parameterized, {PolymorphicEmbed, _options}}} -> true
        _ -> false
      end
    end)
  end
```

atdrendel replied:
👍

fel-mazo commented:
is this abandoned ? I can pick it up if you want

atdrendel committed:
Support current polymorphic embeds

- Detect current and legacy PolymorphicEmbed parameterized types, including arrays
- Pass insert options through to the repo
- Add strategy tests for polymorphic embeds and insert options

atdrendel committed:
Update dependencies

atdrendel committed:
Handle missing repo in polymorphic embed inserts

atdrendel committed:
Address pull request comments

atdrendel commented on lib/ex_machina/ecto_polymorphic_embed_strategy.ex:
```diff
@@ -0,0 +1,176 @@
+    raise "expected :repo to be given to ExMachina.EctoPolymorphicEmbedStrategy"
+  end
+
+  def handle_insert(_, %{repo: nil}, _insert_options) do
+    raise """
+    insert/1 is not available unless you provide the :repo option. Example:
+
+    use ExMachina.EctoPolymorphicEmbed, repo: MyApp.Repo
```
Done.

atdrendel commented on lib/ex_machina/ecto_polymorphic_embed.ex:
```diff
@@ -0,0 +1,380 @@
+  nice things that make working with Ecto easier.
+
+  * It uses `ExMachina.EctoPolymorphicEmbedStrategy`, which adds `insert/1`, `insert/2`,
+  `insert/3` `insert_pair/2`, `insert_list/3`.
```
Done.

atdrendel commented on test/ex_machina/ecto_polymorphic_embed_strategy_test.exs:
```diff
@@ -0,0 +1,118 @@
+  end
+
+  test "insert/3 raises helpful error message if no repo is provided" do
+    assert_raise RuntimeError, ~r/insert\/1 is not available unless you provide the :repo option/, fn ->
```
Um...OK. Done.

atdrendel commented on lib/ex_machina/ecto_polymorphic_embed_strategy.ex:
```diff
@@ -0,0 +1,168 @@
+    raise "expected :repo to be given to ExMachina.EctoPolymorphicEmbedStrategy"
+  end
+
+  def handle_insert(
+        %{__meta__: %{__struct__: Ecto.Schema.Metadata}} = record,
+        %{repo: repo},
+        insert_options
+      ) do
+    record
+    |> cast()
+    |> repo.insert!(insert_options)
+  end
```
Done.

atdrendel commented:
```sh
$ mix test
The database for ExMachina.TestRepo has been dropped
The database for ExMachina.TestRepo has been created
Running ExUnit with seed: 807435, max_cases: 16

........................................................................................
Finished in 0.2 seconds (0.1s async, 0.1s sync)
88 tests, 0 failures
```
